### ISSUE TITLE
High CPU usage on logout when using remote Infinispan only setup

### ISSUE THREAD

**Issue by ahus1** [keycloak/keycloak#32270]
### Description

TL;DR: After the change of https://github.com/keycloak/keycloak/pull/31905 we have an Infinispan query running on each logout. Now the nightly run shows the following behavior:

* an increased 50 perc latency on logout of 2+ second, but also other requests (possibly due to contention). 99 perc goes up to 8 seconds.
* very high CPU usage on Infinispan (from 15% to 850% percent CPU usage)
* There were 10_000 entries with a single owner in the session cache only (due to a misconfiguration in our setup, possibly originating from the persistent sessions re-configuration) - https://github.com/keycloak/keycloak-benchmark/issues/934
* Once the number of entries dropped to 0 again during the test, the CPU usage decreased.

(can be moved to the keycloak repository as needed) 

### Motivation

We need to have a fast logout. Before, the map existed in the user session. Not we need a fast query if we want to stay with a query. 

Further analysis is needed to see what Infinispan is doing. GC should be analysed on Infinispan.

These are a screenshots of a nightly run: The two quarter are not logging out, the second two quarters are logging out. During the third quarter, there are still client sessions in the cache. In the fourth quarter, they have expired. 

In the third quarter, Keycloak is slow.

![image](https://github.com/user-attachments/assets/13d28e75-7752-4fbf-a058-a1b7c01bfc85)

Number of user sessions goes down, but actually quite a lot later than the results return to normal response times:

![image](https://github.com/user-attachments/assets/5c13ede3-e739-4adb-9d7a-be38b8d64da4)

Number of client sessions don't expire on the same speed. This is strange it its own way. 

![image](https://github.com/user-attachments/assets/d90eeb43-1571-4ae1-84ba-68204b74eb13)

**Comment by ahus1** [comment]
To make the nightly runs in KCB work again, the run with external Infinispan only and no persistent sessions has been disabled, see https://github.com/keycloak/keycloak-benchmark/commit/e04188d1c7768ca038009683bdbcc9b2aad67cf4

**Comment by ahus1** [comment]
While #32590 gave a significant improvement, we still see a high CPU usage in Infinispan. Indexing in Infinispan might resolve this, and therefore indexing might be the next step.

**Comment by pruivo** [comment]
### Infinispan Indexing results

I've compiled Infinispan with the ISPN-16473 fix to check the behavior with indexing enabled and the results are promising.

Below is the Grafana data from Keycloak server
![infinispan-histogram](https://github.com/user-attachments/assets/d82f0963-d121-4428-be51-bd6e61944444)

And here are the Gatling reported results
![infinispan](https://github.com/user-attachments/assets/543355ad-9885-4714-926b-f7bad36511f3)

The results are with 100% logout rate and with a hot cache (3 nodes, ~30k unique sessions per node, and ~60k stored (`owners=2`). A total of almost ~100K unique sessions in the cluster).

When compared with persistent sessions, Infinispan achieves similar results. For the record, below is the persistent sessions results from the nightly build:

![aurora](https://github.com/user-attachments/assets/0e4ef5ac-5ef4-4e27-80a7-e4733e595a80)

**Comment by pruivo** [comment]
### Non-Indexed

I couldn't make non-indexed queries to perform any good. At the end of the day, everything I tried took double-digit seconds to execute (~30k unique sessions per node).

**Comment by pruivo** [comment]
### Future work: removal of `ORDER BY` (distributed sorting)

Future work would be to investigate the cost of distributed sorting. It is required when using pagination and the current code uses. It fetches 1000 client sessions per request. 
It would be great to have an estimation of how many client sessions are associated with a single user session to fine-tune this number. 

Also, it may be more performant to fetch everything in a single request. Internally, the `UserSession` exposes everything in a `Map` of client-ids and client sessions, so pagination may not bring any benefit at all, independently of the number of client sessions.

**Comment by ahus1** [comment]
Removing the KC26 milestone, as the RHDG will not arrive in time for RHBK26.

**Comment by pruivo** [comment]
Performance issue fixed in Infinispan 15.0.9.Final.

Today's results:

| Metric | Persistent Sessions | Volatile Sessions |
| --- | --- | --- |
| `userLoginsPerSecPer1vCpuPerPod` | 94 | 97  |
| `meanResponseTime` | 92ms | 68ms |
